### ISSUE TITLE
[survey] DiaASQ

### ISSUE THREAD

**Issue by Nana2929** [Nana2929/dialogue-absa#7]
### 2023 paper
- paper itself is the SOTA
- github: https://github.com/unikcc/DiaASQ 
### Method summary 
1. base encoder learns base contextual repr., 
2. multi-view interaction layers use 3 feature masks (Thread mask, Speaker mask, Reply Mask) + max pooling over masked attns 
3. RoPE 
4. Grid-tagging
![image](https://github.com/Nana2929/sentiment-llm/assets/58811089/b506f790-3384-4d9f-9d9e-9ff04379544a)
 
### Results
- All f1 are strict f1 (spans and elements need to be completely correct) 
![image](https://github.com/Nana2929/sentiment-llm/assets/58811089/28e0aca2-04aa-4e37-a5dc-fd3f049d1cd2)

### Ablation 
- removing all feature masks
![image](https://github.com/Nana2929/sentiment-llm/assets/58811089/a4e348ca-60b1-4d46-b96f-e43d9245cbb7)

**Comment by Nana2929** [comment]
## Code
- 3 masks (3 attentions)
```python
def build_attention(self, sequence_outputs, speaker_masks=None, reply_masks=None, thread_masks=None):
        """
        sequence_outputs: batch_size, seq_len, hidden_size
        speaker_matrix: batch_size, num, num 
        head_matrix: batch_size, num, num 
        """
        speaker_masks = speaker_masks.bool().unsqueeze(1)
        reply_masks = reply_masks.bool().unsqueeze(1)
        thread_masks = thread_masks.bool().unsqueeze(1)

        rep = self.reply_attention(sequence_outputs, sequence_outputs, sequence_outputs, reply_masks)[0]
        thr = self.thread_attention(sequence_outputs, sequence_outputs, sequence_outputs, thread_masks)[0]
        sp = self.speaker_attention(sequence_outputs, sequence_outputs, sequence_outputs, speaker_masks)[0]

        r = torch.stack((rep, thr, sp), 0)
        r = torch.max(r, 0)[0]
        return r
``` 
- custom forward() 
```python
    def forward(self, **kwargs):
        input_ids, input_masks, input_segments = [kwargs[w] for w in ['input_ids', 'input_masks', 'input_segments']]
        ent_matrix, rel_matrix, pol_matrix = [kwargs[w] for w in ['ent_matrix', 'rel_matrix', 'pol_matrix']]
        reply_masks, speaker_masks, thread_masks = [kwargs[w] for w in ['reply_masks', 'speaker_masks', 'thread_masks']]
        sentence_masks, full_masks, dialogue_length = [kwargs[w] for w in ['sentence_masks', 'full_masks', 'dialogue_length']]

        sequence_outputs = self.bert(input_ids, token_type_ids=input_segments, attention_mask=input_masks)[0]

        sequence_outputs =  self.merge_sentence(sequence_outputs, input_masks, dialogue_length)
        sequence_outputs = self.dropout(sequence_outputs)

        sequence_outputs = self.build_attention(sequence_outputs, reply_masks=reply_masks, speaker_masks=speaker_masks, thread_masks=thread_masks)

        loss0, tags0 = self.classify_matrix(kwargs, sequence_outputs, ent_matrix, sentence_masks, 'ent')
        loss1, tags1 = self.classify_matrix(kwargs, sequence_outputs, rel_matrix, full_masks, 'rel')
        loss2, tags2 = self.classify_matrix(kwargs, sequence_outputs, pol_matrix, full_masks, 'pol')
      
        return (loss0, loss1, loss2), (tags0, tags1, tags2)
```
